### PR TITLE
Add coverage tracking via codecov.io

### PR DESCRIPTION
**Pull request by sublimator**
See: https://travis-ci.org/sublimator/rippled/builds/79101415

### CODE REVIEW

**Comment by sublimator** [comment]
See: https://codecov.io/github/sublimator/rippled?branch=codecov.io.4 (or if that is 404 then https://codecov.io/github/sublimator/rippled?branch=codecov.io.3)

Also see chrome extension: https://chrome.google.com/webstore/detail/codecov-extension/keefkhehidemnokodkdkejapdgfjmijf

File: bin/ci/ubuntu/install-dependencies.sh [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+sudo apt-get -y install gdb
+# What versions are we ACTUALLY running?
+g++ -v
+clang -v
+# Avoid `spurious errors` caused by ~/.npm permission issues
+# Does it already exist? Who owns? What permissions?
+ls -lah ~/.npm || mkdir ~/.npm
+# Make sure we own it
+sudo chown -R $USER ~/.npm
+# We use this so we can filter the subtrees from our coverage report
```
Comment (sublimator) [inline bin/ci/ubuntu/install-dependencies.sh]: Also, the bash script seems to fall apart with the amount of source files we have so as to take upwards of 45 minutes for a coverage TARGET build.

Comment (vinniefalco) [inline bin/ci/ubuntu/install-dependencies.sh]: What do you mean "fall apart?"

Comment (sublimator) [inline bin/ci/ubuntu/install-dependencies.sh]: It just performs so badly under the weight of so many files, that that
imagery was warranted. The build length blowout was mentioned after iirc.
Anyway, I can reword it to whatever you like.

Comment (vinniefalco) [inline bin/ci/ubuntu/install-dependencies.sh]: Is it something that can be fixed? Why does the bash script need to manipulate a lot of files? We don't have THAT many sources...

File: SConstruct [hunk 2]
```diff
@@ -15,11 +15,13 @@
 
     clang           All clang variants
     clang.debug     clang debug variant
+    clang.debug     clang coverage variant
```
Comment (vinniefalco) [inline SConstruct]: They both have the same name?

File: bin/ci/ubuntu/install-dependencies.sh (hunk 1 above)
Comment (sublimator) [inline bin/ci/ubuntu/install-dependencies.sh]: > Is it something that can be fixed?

Yeah, by using their Python script, reporting the bash performances issues,
and using a fork of the Python script until they roll in support for our
requirements.

Done. Done. Done.

File: SConstruct (hunk 2 above)
Comment (sublimator) [inline SConstruct]: Yeah, just a sloppy/paste err

**Comment by sublimator** [comment]
@ximinez @vinniefalco

File: .travis.yml [hunk 3]
```diff
@@ -3,68 +3,25 @@ compiler:
...
-  # What versions are we ACTUALLY running?
-  - g++ -v
-  - clang -v
-  # Avoid `spurious errors` caused by ~/.npm permission issues
-  # Does it already exist? Who owns? What permissions?
-  - ls -lah ~/.npm || mkdir ~/.npm
-  # Make sure we own it
-  - sudo chown -R $USER ~/.npm
-
+  - bin/ci/ubuntu/install-dependencies.sh
```
Comment (ximinez) [inline .travis.yml]: I suggest you roll back the `install-dependencies.sh` because #1279 converts all but the last ~4 commands to Travis config.

Comment (sublimator) [inline .travis.yml]: For #1279, are we going to wait until the caching is ready?

Comment (ximinez) [inline .travis.yml]: Travis containers already have caching that we are taking advantage of with those changes.

Comment (sublimator) [inline .travis.yml]: Ah, I thought I read somewhere summing about the build times taking ages,
but will be fixed when caching summin summin

Comment (sublimator) [inline .travis.yml]: Anyway, this P.R. predates that P.R. so we should merge this in cause
coverage is long overdue :) And the chrome extension is bosss. Also, that
chrome extension is boss.

Comment (ximinez) [inline .travis.yml]: To clarify: The first build of the PR was taking ages, because they were building boost. However, later builds will take advantage of the PR's cache, and from what I can tell, once it's merged up far enough (develop or master, I'm not sure which), that branch's cache will be offered to every other build, so we'll have a significant speed up. When we move to the next version of boost, we'll repeat the process, but by then we should know what to expect.

Comment (ximinez) [inline .travis.yml]: > Anyway, this P.R. predates that P.R. so we should merge this in cause coverage is long overdue :)

That works for me. I can handle the merge on my PR. :+1: 

> And the chrome extension is bosss. Also, that chrome extension is boss.

Firefox extension is pretty, too.

**Comment by ximinez** [comment]
:+1:

File: .travis.yml (hunk 3 above)
Comment (ximinez) [inline .travis.yml]: Oh, hey, you know what would be cool? Add a link to the extensions in an appropriate `.md` file, or somewhere appropriate in the travis.yml. Otherwise, nobody's going to know it's there, and nobody's going to use it once this PR is closed and forgotten.

Comment (sublimator) [inline .travis.yml]: Oh, I think we will be using it heaps, cause we can look at new code in pull wrecks and see if it's covered. 

I hope this is merged in ASAP because it will be useful for all subsequent pull requests.

File: bin/ci/ubuntu/install-dependencies.sh (hunk 1 above)
Comment (sublimator) [inline bin/ci/ubuntu/install-dependencies.sh]: And we DO have THAT many source files, when you consider the subtrees. For the moment, I filtered all except beast/ripple.

File: bin/ci/ubuntu/install-dependencies.sh [hunk 4]
```diff
@@ -0,0 +1,32 @@
+#!/bin/bash
+sudo apt-get update -qq
+sudo apt-get install -qq python-software-properties
+sudo add-apt-repository -y ppa:ubuntu-toolchain-r/test
+sudo add-apt-repository -y ppa:afrank/boost
```
Comment (sublimator) [inline bin/ci/ubuntu/install-dependencies.sh]: Still with the `ppa:afrank/boost`

Comment (ximinez) [inline bin/ci/ubuntu/install-dependencies.sh]: @sublimator Don't stress about that too much. It goes away with #1279.

**Comment by sublimator** [comment]
Should I squash this down?

**Comment by ximinez** [comment]
@sublimator May as well.

**Comment by sublimator** [comment]
This passed. The failure for the `debug.nounity` was just a travis fart:
```
Your test run exceeded 50 minutes. 
One possible solution is to split up your test run.
```

**Comment by ximinez** [comment]
I kicked the spurious `debug.nounity` Travis build failure and it succeed this time.

**Comment by ximinez** [comment]
The changes from this PR have been merged into #1279 and conflicts resolved.
